### DISCUSSION TITLE
Unwanted system libraries in custom vim Formula

### DISCUSSION THREAD

**Discussion by PatTheMav** [Homebrew/discussions#735] [Linux] [answered]
Ever since Homebrew removed configurable Formulas I've been running a custom tap that adds back this functionality and for macOS this works absolutely fine.

Recently my test-bot workflow on Linux always fail with the following error for vim:

```
==> brew linkage --test patthemav/custom/vim-custom
==> FAILED
Unwanted system libraries:
  /lib/x86_64-linux-gnu/libselinux.so.1
```

My vim-custom formula is based on the original vim formula before optional features were deprecated and kept up with the changes to the official formula. While the macOS workflow finishes without issues, the Linux workflow acts up.

I have a hunch that I might be missing a dependency on Linux that would otherwise be installed by either Ruby, Perl, Python or Lua (which are _not_ optional in the default formula) and results in vim be linked against system-provided `libselinux`.

As I have no experience with Linuxbrew except for the brew-bot workflow runs, I'd love to hear suggestions how to fix this (e.g. adding a Linux-only dependency).

**Comment by tlk** [comment]
Unless you want the formula to work on linux, consider disabling the test in .github/workflows/tests.yml:
```diff
-        os: [ubuntu-latest, macOS-latest]
+        os: [macOS-latest]
```

**Reply by PatTheMav** [reply]
Yeah that'd my "last resort" option to avoid merging PRs with failed jobs. Right now I'd still like it to work with the templates generated by brew itself and better understand where that unwanted linked library might come from.

**Reply by tlk** [reply]
I'm not an expert, but maybe `libselinux.so.1` needs to be added to SYSTEM_LIBRARY_ALLOWLIST: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/linux/linkage_checker.rb

See also https://github.com/Homebrew/brew/blob/master/Library/Homebrew/linkage_checker.rb#L48

**Reply by tlk** [reply]
Hmm, perhaps the ubuntu-latest image has changed recently?

Selinux in github virtual environments was mentioned in this issue that was closed five days ago:
  https://github.com/actions/virtual-environments/issues/2307

Selinux was mentioned in this formula update three days ago:
  https://github.com/Homebrew/homebrew-core/pull/70421

I just checked and libselinux.so.1 is present today on the ubuntu-latest virtual environment:
  https://github.com/tlk/github-actions-ubuntu-selinux/runs/1850469673

```
libselinux1/bionic,now 2.7-2build2 amd64 [installed,automatic]
```

**Reply by PatTheMav** [reply]
Interesting - fwiw I also had brew-bot complain about lots of other "unwanted" libraries before which was fixed by removing the optional x11 integration. I haven't dug too deep into this, but it seems that brew-bot builds with all optional dependencies enabled maybe?

**Comment by carlocab** [comment, marked as answer]
You could try passing a `--without-selinux` flag to `./configure`.

**Reply by PatTheMav** [reply]
Yep I've discovered that flag in the configure script as well, probably the easiest way to fix this. I wonder if this happens with the official Formula as well or if it just isn't tested/automatically discovered at this time.

**Comment by PatTheMav** [comment]
Quick update: I updated my vim formula today and added the `--without-selinux` flag on Linux, but `brew linkage` still fails and complains about it.

**Reply by danielnachun** [reply]
Try `--disable-selinux` instead.  From https://github.com/vim/vim/blob/7d41410a455ba62128655beb91c35d3fba299287/src/configure.ac#L478

**Reply by PatTheMav** [reply]
Of course.. 🤦🏻‍♂️

**Reply by danielnachun** [reply]
I had to do this for a bunch of formula to make them build on Linux and it was a complete coin toss whether it would be `--without-selinux` or `--disable-selinux`.  Not sure why developers couldn't be more consistent for this but it's super annoying!

**Reply by PatTheMav** [reply]
Thanks, that did the trick.

**Comment by thiagogenez** [comment]
Many thanks for the flag `--disable-selinux`. It worked here.  I hit this error a few days ago and it was annoying.

**Reply by danielnachun** [reply]
If you find any formulae which need this to be added feel free to make a PR.  I was able to catch a lot of them when I first set up Homebrew on CentOS, but there are probably still packages out there which need it and it can be very difficult to track down.

**Reply by thiagogenez** [reply]
Thanks, @danielnachun. I have recently set up Homebrew on Centos because of the working environment. Happy to contribute to investigating these issues as soon as I catch them, and make a PR to fixing it on CentOS.
